server/src/main/java/org/opensearch/common/bytes/AbstractBytesReference.java

```diff
@@ -82,7 +84,13 @@ public void writeTo(OutputStream os) throws IOException {
 
     @Override
     public String utf8ToString() {
-        return toBytesRef().utf8ToString();
+        BytesRef bytesRef = toBytesRef();
+        final char[] ref = new char[bytesRef.length];
+        final int len = UnicodeUtil.UTF8toUTF16(bytesRef, ref);
```

Comment:
That is the max length to ensure that ref's length is large enough to ensure the conversion doesn't overflow. Not sure if I got your comment correctly. Are you saying that we can optimise `UTF8toUTF16` computation? If yes, that could probably be opened up with Lucene

Reply:
I am thinking if it makes sense to introduce `calcUTF8toUTF16Length` method (similar to `calcUTF16toUTF8Length`) to estimate the length of the conversion without allocation? Also could be easier to test without triggering OOMs...

Reply:
Wouldn't it add an additional overhead where in we calculate the length always first?

Reply:
It may add for sure (we could measure the overhead with `jmh`), ideally we would need a method which does conversion but stops when "limit" reaches (we don't have one sadly). Also, we could be pretty smart and consult `bytesRef.length` before only when we suspect the overlimits:
```
if (bytesRef.length > MAX_LENGTH) {
    final int len = UnicodeUtil.calcUTF8toUTF16Length (bytesRef);
    ...
}
```